data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/codec/CSVCodec.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugins.source.codec;
+
+import com.amazon.dataprepper.model.annotations.DataPrepperPlugin;
+import com.amazon.dataprepper.model.annotations.DataPrepperPluginConstructor;
+import com.amazon.dataprepper.model.event.Event;
+import com.amazon.dataprepper.model.log.JacksonLog;
+import com.amazon.dataprepper.model.record.Record;
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.databind.MappingIterator;
+import com.fasterxml.jackson.dataformat.csv.CsvMapper;
+import com.fasterxml.jackson.dataformat.csv.CsvParser;
+import com.fasterxml.jackson.dataformat.csv.CsvReadException;
+import com.fasterxml.jackson.dataformat.csv.CsvSchema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+@DataPrepperPlugin(name = "csv", pluginType = Codec.class, pluginConfigurationType = CSVCodecConfig.class)
+public class CSVCodec implements Codec {
+    private final CSVCodecConfig config;
+    private static final Logger LOG = LoggerFactory.getLogger(CSVCodec.class);
+
+    @DataPrepperPluginConstructor
+    public CSVCodec(final CSVCodecConfig config) {
+        Objects.requireNonNull(config);
+        this.config = config;
+    }
+
+    @Override
+    public void parse(final InputStream inputStream, final Consumer<Record<Event>> eventConsumer) throws IOException {
+        try (final BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
+            parseBufferedReader(reader, eventConsumer);
+        }
+    }
+
+    private void parseBufferedReader(final BufferedReader reader, final Consumer<Record<Event>> eventConsumer) throws IOException {
+        final CsvMapper mapper = createCsvMapper();
+        final CsvSchema schema;
+        if (config.isDetectHeader()) {
+            // autodetect header from the first line of csv file
+            schema = createAutodetectCsvSchema();
+        }
+        else {
+            // construct a header from the pipeline config or autogenerate it
+            final int defaultBufferSize = 8192; // number of chars before mark is ignored (this is the buffer size, so large header files
+            // can be read since more buffers will be allocated.)
+            reader.mark(defaultBufferSize); // getting number of columns of first line will consume the line, so mark initial location
+
+            final int firstLineSize = getSizeOfFirstLine(reader.readLine());
+            reader.reset(); // move reader pointer back to beginning of file in order to parse first line
+            schema = createCsvSchemaFromConfig(firstLineSize);
```

Comment:
If the first line of the reader contains the header do we need when creating the MappingIterator in the section below? Is the `mark` and `reset` stuff necessary?

Reply:
The branch in the `else` statement is triggered if there's no header in the csv file. So we want to `mark` and `reset` so that the first row of the file can be parsed into `parsingIterator` a second time (after it's used to figure out the actual number of columns in the csv file, and generate more column names for the header if necessary). I moved this logic to a helper method with a descriptive name which should hopefully improve readability.